llvm/lib/Transforms/InstCombine/InstCombineSelect.cpp

```diff
@@ -4638,5 +4681,20 @@ Instruction *InstCombinerImpl::visitSelectInst(SelectInst &SI) {
                                      Align(MaskedLoadAlignment->getZExtValue()),
                                      CondVal, FalseVal));
 
+  // Canonicalize sign function ashr pattern: select (icmp slt X, 1), ashr X,
```

Comment:
Please file a separate patch for this change.
We can also fold the commuted pattern:
```
select (icmp sgt X, 0), 1, (ashr X, BW - 1) -> scmp(X, 0)
```